packages/terra-functional-testing/lib/services/wdio-docker-service.js

```diff
@@ -0,0 +1,167 @@
+
+    if (stackInfo) {
+      logger.log('Removing docker stack.');
+
+      await exec('docker stack rm wdio');
+
+      // Ensure the services and network have been removed.
+      await this.awaitServiceRemoval();
+      await this.awaitNetworkRemoval();
+    }
+  }
+
+  /**
+   * Waits for a command to complete successfully.
+   * @param {string} command - The shell command to run.
+   * @param {func} callback - A callback function to accept or reject the result of the command. Must return a promise.
+   * @param {number} retries - The number of times to retry the command. Defaults to 30 retries.
+   * @param {number} interval - The timeout between commands in milliseconds. Defaults to every two seconds.
+   * @returns {Promise} - A promise that resolves when the callback accepts the command response.
+   */
+  async pollCommand(command, callback, retries = RETRY_COUNT, interval = POLL_INTERVAL) {
+    return new Promise((resolve, reject) => {
+      let retryCount = 0;
+      let pollTimeout = null;
+
+      const poll = async () => {
+        if (retryCount >= retries) {
+          clearTimeout(pollTimeout);
+          pollTimeout = null;
+          reject(Error(logger.format('Timeout. Exceeded retry count.')));
+        }
+
+        try {
+          const result = await exec(command);
+
+          await callback(result).then(() => resolve());
+        } catch (error) {
+          retryCount += 1;
+          pollTimeout = setTimeout(poll, interval);
+        }
+      };
+
+      pollTimeout = setTimeout(poll, interval);
+    });
+  }
+
+  /**
+   * Ensures the docker network has been shut down.
+   */
+  async awaitNetworkRemoval() {
+    await this.pollCommand('docker network ls | grep wdio || true', (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout: networkStatus } = result;
+
+        // Reject if there is an active network returned.
+        if (networkStatus) {
+          reject();
+        } else {
+          resolve();
+        }
+      })));
+  }
+
+  /**
+   * Ensures the docker services have been shut down.
+   */
+  async awaitServiceRemoval() {
+    await this.pollCommand('docker service ls | grep wdio || true', (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout: serviceStatus } = result;
+
+        // Reject if there is an active service returned.
+        if (serviceStatus) {
+          reject();
+        } else {
+          resolve();
+        }
+      })));
+  }
+
+  /**
+   * Ensures the docker network is ready.
+   */
+  async awaitNetworkReady() {
+    logger.log('Waiting for docker to become ready.');
+
+    await this.pollCommand(`curl -sSL http://${this.host}:${this.port}/wd/hub/status`, (result) => (
+      new Promise((resolve, reject) => {
+        const { stdout } = result;
+        const { value } = JSON.parse(stdout);
+
+        if (value.ready) {
+          resolve();
+        } else {
+          reject();
```

Comment:
Can we add a log for when the grid is not ready & it causes the service to fail the onPrepare setup?

Reply:
If this is rejected or exceeds the retry count the timeout message will bubble and be logged to the console:

```
2020-06-02T21:22:49.364Z INFO @wdio/cli:launcher: Run onPrepare hook
[terra-functional-testing:wdio-docker-service] Deploying docker stack.
[terra-functional-testing:wdio-docker-service] Waiting for docker to become ready.
2020-06-02T21:22:58.681Z ERROR @wdio/cli:utils: A service failed in the 'onPrepare' hook
Error: [terra-functional-testing:wdio-docker-service] Timeout. Exceeded retry count.
    at Timeout.poll (/Users/se028266/workspaces/terra/terra-toolkit/packages/terra-functional-testing/lib/services/wdio-docker-service.js:86:18)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)

Continue...
```

We can catch the error and bubble a more descriptive one if we'd like. The timeout message is generic since it's used for different things